### PR TITLE
Use libvmi in tests/storage

### PR DESCRIPTION
**Pull request by 0xFelix**
**What this PR does / why we need it**:

This PR introduces new functions to libvmi and
changes the tests in tests/storage to use libvmi.

**Release note**:

```release-note
None
```

### CODE REVIEW

File: tests/libvmi/storage.go [hunk 1]
```diff
@@ -79,6 +95,12 @@ func addVolume(vmi *v1.VirtualMachineInstance, volume v1.Volume) {
 	}
 }
 
+func addFilesystem(vmi *v1.VirtualMachineInstance, filesystem v1.Filesystem) {
+	if !filesystemExists(vmi, filesystem) {
```
Comment (brybacki) [inline tests/libvmi/storage.go:99]: I see this here, and in many other places, we just silently ignore the problem, maybe it would be better to error or panic here if someone adds a Volume or Disk and some objects already exist?

But I suppose that is out of scope for this PR. But we should definitely discuss it.

File: tests/storage/storage.go [hunk 2]
```diff
@@ -912,8 +912,6 @@ var _ = SIGDescribe("Storage", func() {
 							libvmi.WithResourceMemory("256Mi"),
 							libvmi.WithNetwork(v1.DefaultPodNetwork()),
 							libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()))
-
-						vmi = tests.NewRandomVMIWithPVC(fmt.Sprintf("disk-%s", pvc))
```
Comment (brybacki) [inline tests/storage/storage.go:916]: Good catch, that was definitely an error left after previous refactoring

File: tests/nonroot_test.go [hunk 3]
```diff
@@ -36,7 +38,7 @@ var _ = Describe("[sig-compute]NonRoot feature", func() {
 				Skip(fmt.Sprintf("Missing %s, enable %s featureGate.", virtconfig.VirtIOFSGate, virtconfig.VirtIOFSGate))
 			}
 
-			vmi := tests.NewRandomVMIWithPVCFS("test")
+			vmi := libvmi.NewFedora(libvmi.WithFilesystemPVC("virtiofs", "test"))
```
Comment (brybacki) [inline tests/nonroot_test.go:41]: Why fedora?
It probably does not make a difference here.

File: tests/storage/storage.go [hunk 4]
```diff
@@ -476,28 +465,27 @@ var _ = SIGDescribe("Storage", func() {
...
 				virtiofsTestFile := fmt.Sprintf("%s/virtiofs_test", virtiofsMountPath)
 				mountVirtiofsCommands := fmt.Sprintf(`#!/bin/bash
                                        mkdir %s
                                        mount -t virtiofs %s %s
                                        touch %s
-                               `, virtiofsMountPath, fs.Name, virtiofsMountPath, virtiofsTestFile)
-				tests.AddUserData(vmi, cloudInitName, mountVirtiofsCommands)
+                               `, virtiofsMountPath, fsName, virtiofsMountPath, virtiofsTestFile)
 
+				vmi := libvmi.NewFedora(libvmi.WithFilesystemDV("disk1", dataVolume.Name))
```
Comment (brybacki) [inline tests/storage/storage.go:484]: a leftover from previous versions?

File: tests/storage/storage.go [hunk 5]
```diff
@@ -700,7 +688,7 @@ var _ = SIGDescribe("Storage", func() {
 			})
 			It("VirtioFS, it should fail to start a VMI", func() {
 				tests.DisableFeatureGate(virtconfig.VirtIOFSGate)
-				vmi := tests.NewRandomVMIWithFSFromDataVolume("something")
+				vmi := libvmi.NewFedora(libvmi.WithFilesystemDV("disk1", "something"))
```
Comment (brybacki) [inline tests/storage/storage.go:691]: it does not matter here, as any string would work, but might be misleading, as the first argument is a filesystem name not a disk name

File: tests/storage/storage.go [hunk 6]
```diff
@@ -1056,10 +1044,12 @@ var _ = SIGDescribe("Storage", func() {
...
 			It("[test_id:1015]should be successfully started", func() {
 				// Start the VirtualMachineInstance with the PVC attached
-				vmi = tests.NewRandomVMIWithPVC(dataVolume.Name)
 				// Without userdata the hostname isn't set correctly and the login expecter fails...
-				tests.AddUserData(vmi, cloudInitName, "#!/bin/bash\necho 'hello'\n")
-
+				vmi = libvmi.New(
+					libvmi.WithResourceMemory("256Mi"),
+					libvmi.WithPersistentVolumeClaim("disk0", dataVolume.Name, v1.DiskBusVirtio),
+					libvmi.WithCloudInitNoCloudUserData("#!/bin/bash\necho 'hello'\n", true),
```
Comment (brybacki) [inline tests/storage/storage.go:1051]: can you make this string a const and name it cirrosHack? 

And we need to find a way to get rid of this in near future, maybe switch to Alpine? But the switch of OS is an idea for another PR.

File: tests/storage/storage.go [hunk 7]
```diff
@@ -1131,39 +1123,10 @@ var _ = SIGDescribe("Storage", func() {
...
-					Name: "emptydisk2",
-					VolumeSource: virtv1.VolumeSource{
-						EmptyDisk: &virtv1.EmptyDiskSource{
-							Capacity: resource.MustParse("1Gi"),
-						},
-					},
-				})
+				vmi = libvmi.NewAlpine(
+					libvmi.WithEmptyDisk("emptydisk1", "1Gi", v1.DiskBusSCSI),
+					libvmi.WithEmptyDisk("emptydisk2", "1Gi", v1.DiskBusSATA),
```
Comment (brybacki) [inline tests/storage/storage.go:1128]: :+1: 
this looks so good, when compared to previous one ;)

**Comment by brybacki** [comment]
I had just a few comments. Looks good. 

This storage test failed on a few lanes:
```
Tests Suite: [sig-storage] Hotplug [storage-req] VMI migration should allow live migration with attached hotplug volumes expand_less | 2m44s
-- | --
 Failure tests/storage/hotplug.go:934
```

Maybe something does not work the same way as before?

File: tests/libvmi/storage.go (hunk 1 above)
Comment (0xFelix) [reply tests/libvmi/storage.go:99]: You are right, we should discuss this / change it in a separate PR.

File: tests/nonroot_test.go (hunk 3 above)
Comment (0xFelix) [reply tests/nonroot_test.go:41]: It was fedora before but actually it should make no difference. Let's try with `libvmi.New` instead.

File: tests/storage/storage.go (hunk 4 above)
Comment (0xFelix) [reply tests/storage/storage.go:484]: Yep, missed that, thanks!

File: tests/storage/storage.go (hunk 5 above)
Comment (0xFelix) [reply tests/storage/storage.go:691]: Changed it to `fs0`

File: tests/storage/storage.go (hunk 7 above)
Comment (0xFelix) [reply tests/storage/storage.go:1128]: Thanks :)

File: tests/storage/hotplug.go [hunk 8]
```diff
@@ -612,13 +604,13 @@ var _ = SIGDescribe("Hotplug", func() {
 					Skip("Skip test when RWXBlock storage class is not present")
 				}
 
-				template := libvmi.NewCirros()
+				vmi := libvmi.NewCirros()
 				node := findCPUManagerWorkerNode()
 				if node != "" {
-					template.Spec.NodeSelector = make(map[string]string)
-					template.Spec.NodeSelector[corev1.LabelHostname] = node
+					vmi.Spec.NodeSelector = make(map[string]string)
```
Comment (dankenigsberg) [inline tests/storage/hotplug.go:610]: In a follow-up PR you may want to use `libvmi.WithNodeSelectorFor`

File: tests/storage/storage.go (hunk 7 above)
Comment (dankenigsberg) [reply tests/storage/storage.go:1128]: I'd say that it's so good that it makes the comment above it redundant.

File: tests/nonroot_test.go (hunk 3 above)
Comment (brybacki) [reply tests/nonroot_test.go]: If it was fedora, maybe it needed guest agent? I would not change it in a PR that is supposed to be a refactoring. unless you are sure it will be much better.

Comment (0xFelix) [reply tests/nonroot_test.go]: The test wants the creation of the VM to fail with a specific error message so presence of the guest agent should not matter.

File: tests/storage/storage.go [hunk 9]
```diff
@@ -1314,24 +1277,23 @@ var _ = SIGDescribe("Storage", func() {
...
-				tests.AddPVCDisk(vmi2, diskName, v1.DiskBusVirtio, pvcClaim)
-
 				setShareable(vmi1, diskName)
 				setShareable(vmi2, diskName)
 
 				By("Starting the VirtualMachineInstances")
-				tests.RunVMIAndExpectLaunch(vmi1, 500)
-				tests.RunVMIAndExpectLaunch(vmi2, 500)
+				vmi1 = tests.RunVMIAndExpectLaunch(vmi1, 500)
+				vmi2 = tests.RunVMIAndExpectLaunch(vmi2, 500)
```
Comment (EdDev) [inline tests/storage/storage.go:1296]: Not in this PR, but note that the test time can be improved by starting both in parallel and waiting for both serially.
i.e.:
```
create(vmi1)
create(vmi2)
wait(vmi1)
wait(vmi2)
```

**Review by EdDev — COMMENTED** [review]
Very nice PR!

I had some comments in line.

It is a nit, but it will help to read the log if the commit message subject has a consistent format.
I usually like it to be: `<topic>: <subject>`
In case you touch `libvmi`, then having the topic a `tests, libvmi: ` would allow future readers to classify it as tests relates.

**Review by EdDev — APPROVED** [review]
Thank you!

File: tests/libvmi/storage.go [hunk 10]
```diff
@@ -71,6 +95,12 @@ func addVolume(vmi *v1.VirtualMachineInstance, volume v1.Volume) {
...
```
Comment (xpivarc) [inline tests/libvmi/storage.go:99]: I dislike this as it gives a false feeling that I constructed my VM as I wanted and didn't make mistakes.

Comment (0xFelix) [reply tests/libvmi/storage.go:99]: @brybacki Also mentioned this above. I implemented it like the other `addXYZ` functions for now but it should be changed in another PR.

Comment (xpivarc) [reply tests/libvmi/storage.go:99]: I am lacking reasoning why it should be changed in another PR. To me, this is similar to saying: Let me introduce a bug because there is a similar bug already and hopefully someone will fix both of them in x days.

Comment (0xFelix) [reply tests/libvmi/storage.go]: I intended to be that someone who will fix it in a follow-up PR but might as well do it for `addFilesystem` in this PR. See also https://github.com/kubevirt/kubevirt/pull/8348 for the other ones.

Comment (brybacki) [reply tests/libvmi/storage.go]: better to fix the others in a new PR as we might discover some problems

**Comment by enp0s3** [comment]
/hold
Giving time to address @xpivarc comments

File: tests/libvmi/storage.go [hunk 11]
```diff
@@ -59,6 +59,14 @@ func WithEmptyDisk(diskName, capacity string, bus v1.DiskBus) Option {
 	}
 }
 
+// WithCDRom specifies a CDRom drive backed by a PVC to be used.
+func WithCDRom(cdRomName, claimName string, bus v1.DiskBus) Option {
```
[the remark this answers is not in the document]
Comment (0xFelix) [reply tests/libvmi/storage.go:65]: It is used [here](https://github.com/kubevirt/kubevirt/pull/8300/commits/2f23f678e599eaa233c7deddcd6f8bf82b2bb5ec#diff-5a0009fa8acc3c85167978b2c7d71276f1e8f127067038836c2c36b74fadb31cR267).

**Comment by 0xFelix** [comment]
@xpivarc Addressed your comments, WDYT?

**Review by EdDev — APPROVED** [review]
Looks great, thanks!

Placing a hold to give @xpivarc a chance to feedback on his requests.

/hold

**Comment by xpivarc** [comment]
/hold cancel
My comments were addressed.
